src/java.base/share/classes/java/net/URLEncoder.java

```diff
@@ -239,7 +240,9 @@ public static String encode(String s, Charset charset) {
         }
 
         StringBuilder out = new StringBuilder(s.length() << 1);
-        CharArrayWriter charArrayWriter = new CharArrayWriter();
+        CharsetEncoder ce = charset.newEncoder();
+        CharBuffer cb = CharBuffer.allocate(ENCODING_CHUNK_SIZE);
+        ByteBuffer bb = ByteBuffer.allocate((int)(ENCODING_CHUNK_SIZE * (double)ce.maxBytesPerChar()));
```

Comment:
The cast to (double) is curious

Reply:
It's copied from `String::scale`: 
```
    private static int scale(int len, float expansionFactor) {
        // We need to perform double, not float, arithmetic; otherwise
        // we lose low order bits when len is larger than 2**24.
        return (int)(len * (double)expansionFactor);
    }
```
... though admittedly this caveat doesn't apply since `len` in our case is a small constant value.